File: v2/examples/c/urihandler_v2_test.c
#include "urihandler_v2.h"
#include <assert.h>
#include <string.h>

static int led_called = 0;
static int log_called = 0;

static void led_set(const char* target, const char* const* args, size_t arg_count) {
  assert(strcmp(target, "device-01") == 0);
  assert(arg_count == 1);
  assert(strcmp(args[0], "on") == 0);
  led_called++;
}

static void log_info_user_created(const char* target, const char* const* args, size_t arg_count) {
  assert(strcmp(target, "app") == 0);
  assert(arg_count == 1);
  assert(strcmp(args[0], "42") == 0);
  log_called++;
}

int main(void) {
  uh2_descriptor_t d;
  assert(uh2_parse("device://device-01/led/set/on?trace=1#ui", &d) == 0);
  assert(strcmp(d.package_name, "device") == 0);
  assert(strcmp(d.target, "device-01") == 0);
  assert(d.segment_count == 3);
  assert(strcmp(d.segments[0], "led") == 0);
  assert(strcmp(d.segments[1], "set") == 0);
  assert(strcmp(d.segments[2], "on") == 0);
  assert(uh2_parse("not-a-uri", &d) != 0);
  assert(uh2_parse(NULL, &d) != 0);
  assert(uh2_parse("device://device-01/led", &d) != 0);

  const uh2_route_t routes[] = {
    {"device", "led", "set", led_set},
    {"log", "info", "user-created", log_info_user_created},
  };
  const size_t route_count = sizeof(routes) / sizeof(routes[0]);
  assert(uh2_dispatch("device://device-01/led/set/on", routes, route_count) == 0);
  assert(uh2_dispatch("log://app/info/user-created/42", routes, route_count) == 0);
  assert(uh2_dispatch("device://device-01/motor/set/on", routes, route_count) != 0);
  assert(led_called == 1);
  assert(log_called == 1);
  return 0;
}

File: v2/examples/js/urihandler-v2.test.js
import assert from 'node:assert/strict';
import test from 'node:test';
import {
  dispatch,
  hashUri,
  normalizeUri,
  parseUri,
  resolve,
  translate,
  validate,
} from './urihandler-v2.js';

const registry = {
  device: {
    led: {
      set(target, args, payload, descriptor) {
        return { descriptor, ok: true, payload, state: args[0], target };
      },
    },
  },
  log: {
    info: {
      'user-created'(target, args, payload, descriptor) {
        return { args, descriptor, event: 'user-created', ok: true, payload, sink: target };
      },
    },
  },
};

test('parses, normalizes, and translates logical URI', () => {
  const descriptor = parseUri('device://device-01/led/set/on?trace=1#ui');
  assert.equal(normalizeUri(descriptor), 'device://device-01/led/set/on');
  assert.deepEqual(translate(descriptor).route, ['device', 'led', 'set']);
  assert.deepEqual(translate(descriptor).args, ['device-01', 'on']);
});

test('validates and resolves through route tree', () => {
  const translation = translate(parseUri('device://device-01/led/set/on'));
  const cache = new Map();
  assert.equal(validate(translation, registry), true);
  assert.equal(resolve(translation, registry, cache), registry.device.led.set);
  assert.equal(resolve(translation, registry, cache), registry.device.led.set);
  assert.equal(cache.size, 1);
});

test('dispatches device and log routes', async () => {
  assert.deepEqual(await dispatch('device://device-01/led/set/on', registry, { source: 'test' }), {
    descriptor: {
      fragment: null,
      normalized: 'device://device-01/led/set/on',
      package: 'device',
      query: {},
      raw: 'device://device-01/led/set/on',
      segments: ['led', 'set', 'on'],
      target: 'device-01',
    },
    ok: true,
    payload: { source: 'test' },
    state: 'on',
    target: 'device-01',
  });
  assert.equal((await dispatch('log://app/info/user-created', registry, { userId: 42 })).sink, 'app');
});

test('hashes normalized URI with sha256 hex', () => {
  assert.match(hashUri('device://device-01/led/set/on'), /^[a-f0-9]{64}$/);
});

test('rejects invalid route', async () => {
  await assert.rejects(() => dispatch('device://device-01/motor/set/on', registry), /Route validation failed/);
});
